atecc608a_se.c

```diff
@@ -0,0 +1,224 @@
+    printf("  - Config locked: %d\n", config_locked);
+    ASSERT_STATUS(atcab_is_locked(LOCK_ZONE_DATA, &locked), ATCA_SUCCESS);
+    printf("  - Data locked: %d\n", locked);
+    for(uint8_t i=0; i < 16; i++)
+    {
+        ASSERT_STATUS(atcab_is_slot_locked(i, &locked), ATCA_SUCCESS);
+        printf("  - Slot %d locked: %d\n", i, locked);
+    }
+    ASSERT_STATUS(atcab_release(), ATCA_SUCCESS);
+    printf("--------------------------------\n");
+    if(!config_locked)
+    {
+        printf("Config is not locked! Cannot use private keys.\n");
+        return ATCA_NOT_LOCKED;
+    }
+    return ATCA_CONFIG_ZONE_LOCKED;
+}
+
+psa_status_t atecc608a_export_key(psa_key_slot_number_t key,
+                                         uint8_t *p_data,
+                                         size_t data_size,
+                                         size_t *p_data_length)
+{
+    ATCA_STATUS ret = ATCA_SUCCESS;
+    uint16_t slot = key;
+    size_t key_data_len = 65;
+    if(data_size < key_data_len)
+    {
+        printf("FAILED! Key buffer too small.\n");
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+    ASSERT_STATUS(atcab_init(&atca_iface_config), ATCA_SUCCESS);
+
+    ret = atcab_get_pubkey(slot, &p_data[1]);
+    if (ret != ATCA_SUCCESS)
+    {
+        printf("FAILED! atcab_get_pubkey: %d\n", ret);
+        return atecc608a_to_psa_error(ret);
+    }
+    p_data[0] = 4;
+    *p_data_length = key_data_len;
+    printf("atecc608a_export_key - pubkey size %d:\n", *p_data_length);
+    atcab_printbin_sp(p_data, *p_data_length);
+    ASSERT_STATUS(atcab_release(), ATCA_SUCCESS);
+}
+
+psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
+                                       psa_algorithm_t alg,
+                                       const uint8_t *p_hash,
+                                       size_t hash_length,
+                                       uint8_t *p_signature,
+                                       size_t signature_size,
+                                       size_t *p_signature_length)
+{
+    ATCA_STATUS ret = ATCA_SUCCESS;
+    uint16_t key_id = key_slot;
+
+    /* We can only do ECDSA on SHA-256 */
+    /* PSA_ALG_ECDSA(PSA_ALG_SHA_256) */
+    if(!PSA_ALG_IS_ECDSA(alg))
+    {
+        printf("FAILED! PSA_ALG_IS_ECDSA:\n");
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (hash_length != 32)
+    {
+        printf("FAILED! hash_length:\n");
+        /* The driver only supports signing things of length 32. */
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (!PSA_ALG_IS_RANDOMIZED_ECDSA(alg))
+    {
+        printf("FAILED! PSA_ALG_IS_RANDOMIZED_ECDSA:\n");
+        /* The hardware only supports randomized ECDSA */
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    /* XXX how to check the curve? */
+    //PSA_KEY_TYPE_ECC_KEYPAIR(PSA_ECC_CURVE_SECP256R1);
+    //PSA_KEY_TYPE_ECC_KEYPAIR(PSA_ECC_CURVE_SECP256K1);
```

Comment:
We should check that the curve being requested is what the SE supports (only PSA_ECC_CURVE_SECP256R1).